virtool/uploads/data.py

```diff
@@ -0,0 +1,199 @@
+            uploads.append(result.to_dict())
+
+        uploads = await apply_transforms(uploads, [AttachUserTransform(self._db)])
+
+        return [UploadMinimal(**upload) for upload in uploads]
+
+    async def create(
+        self,
+        name: str,
+        upload_type: str,
+        reserved: bool = False,
+        user: Optional[str] = None,
+    ) -> Upload:
+        """
+        Create an upload.
+        """
+
+        async with AsyncSession(self._pg) as session:
+            upload = Upload(
+                created_at=virtool.utils.timestamp(),
+                name=name,
+                ready=False,
+                removed=False,
+                reserved=reserved,
+                type=upload_type,
+                user=user,
+            )
+
+            session.add(upload)
+
+            await session.flush()
+
+            upload.name_on_disk = f"{upload.id}-{upload.name}"
+
+            upload = upload.to_dict()
+
+            await session.commit()
+
+        return Upload(**upload)
+
+    async def get(self, upload_id: int) -> Optional[Upload]:
+        """
+        Get a single upload by its ID.
+
+        :param upload_id: the upload's ID
+        :return: the upload
+        """
+        async with AsyncSession(self._pg) as session:
+            upload = (
+                await session.execute(
+                    select(Upload).filter_by(id=upload_id, removed=False)
+                )
+            ).scalar()
+
+            if not upload:
+                return None
+
+        return upload
+
+    async def delete(self, upload_id: int) -> Upload:
+        """
+        "Delete" a row in the `uploads` table and remove it from the local disk.
+
+        :param upload_id: Row `id` to "delete"
+        :return: the upload
+        """
+
+        upload = await self.delete_row(upload_id)
+
+        if not upload:
+            return None
+
+        try:
+            await run_in_thread(
+                rm, self._config.data_path / "files" / upload.name_on_disk
+            )
+        except FileNotFoundError:
+            pass
+
+        return upload
+
+    async def delete_row(self, upload_id: int) -> Optional[Upload]:
+        """
+        Set the `removed` and `removed_at` attributes of the upload in the given row.
+
+        :param pg: PostgreSQL AsyncEngine object
+        :param upload_id: Row `id` to set attributes for
+        :return: The upload
+        """
+        async with AsyncSession(self._pg) as session:
+            upload = (
+                await session.execute(select(Upload).where(Upload.id == upload_id))
+            ).scalar()
+
+            if not upload or upload.removed:
+                return None
+
+            upload.reads.clear()
+            upload.removed = True
+            upload.removed_at = virtool.utils.timestamp()
+
+            upload = upload.to_dict()
+
+            await session.commit()
+
+        return Upload(**upload)
+
+    async def finalize(
```

Comment:
This does not abstract the backing services. You have a `model` parameter that is a SQLAlchemy model class.

This method should stay in `db.py`. This 'uploads' piece of the data layer is for user uploads. However, there are other uploads that come from workflows. Both have a finalization step for the SQL row when the upload succeeds.

1. Restore the `virtool.uploads.db.finalize` function.
2. The finalize method of the data layer should just call this function and pass it `Upload` (SQL) for the model argument.